### PR TITLE
[dynamo] Enable typechecking for bytecode_transformation.py

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112565
* #112564
* #112563
* #112562
* __->__ #112561

As part of this diff, I have upgraded the `python_version` config setting to 3.11. `bytecode_transformation.py` (and a few other files) have functions using APIs only available in Python 3.11+. Those APIs are gated by a sys.version_info check in their typeshed .pyi files. So setting the min version to 3.11 allows those functions to typecheck properly.

An alternative is to make the relevant types Any:

```
if sys.version_info >= (3, 11):
    _Positions = dis.Positions
else:
    _Positions = Any
```

However, with python_version = 3.8, that means we're not getting any useful typechecking signal when encountering values of type _Position.

Changing the python_version to 3.11 does mean that we will stop typechecking codepaths that run only on lower versions, but that seems a small price to pay. It does also mean that we won't catch code that uses newer APIs without the appropriate version check, but again, not sure this has much of an impact.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: mypy-nofollow.ini [hunk 1]
```diff
@@ -21,7 +21,7 @@ files =
 
 # Minimum version supported - variable annotations were introduced
 # in Python 3.8
-python_version = 3.8
```
Comment (Skylion007) [inline mypy-nofollow.ini:24]: We still need to support 3.8. You should be able to handle everything with typing_extension backports and not using the latest and flashiest typing syntax.

**Review by Skylion007 — COMMENTED** [review]
Why update mypy?

**Comment by int3** [comment]
> Why update mypy?

This is still a draft PR, I was going to write an explanation :)

Basically bytecode_transformation.py (and a few other files) have functions using APIs only available in Python 3.11+. Those APIs are gated by a `sys.version_info` check in their typeshed .pyi files. So setting the min version to 3.11 allows those functions to typecheck properly.

An alternative is to make the relevant types `Any`:

```
if sys.version_info >= (3, 11):
    _Positions = dis.Positions
else:
    _Positions = Any
```

However, with `python_version = 3.8`, that means we're not getting any useful typechecking signal when encountering values of type `_Position`.

Changing the python_version to 3.11 does mean that we will stop typechecking codepaths that run only on lower versions, but that seems a small price to pay. It does also mean that we won't catch code that uses newer APIs without the appropriate version check, but again, not sure this has much of an impact.

> You should be able to handle everything with typing_extension backports and not using the latest and flashiest typing syntax.

I don't think typing syntax is gated on this config setting. E.g. the `A | B` union syntax is accepted by mypy regardless of what the `python_version` is.

File: torch/_C/_dynamo/eval_frame.pyi [hunk 2]
```diff
@@ -1,7 +1,10 @@
 import types
+from typing import NewType
 
 from torch._dynamo.types import DynamoCallback, DynamoGuardHook
 
+_PyInterpreterFrame = NewType("_PyInterpreterFrame", types.FrameType)
```
Comment (int3) [inline torch/_C/_dynamo/eval_frame.pyi:6]: from what I gather, we implement our own `FrameType`-like type for Python >= 3.11. So it's not actually an alias of FrameType, but still exposes the same interface.

(I will add this as a code comment)

File: mypy-nofollow.ini [hunk 3]
```diff
@@ -21,7 +21,7 @@ files =
...
+python_version = 3.11
```
Comment (Skylion007) [inline mypy-nofollow.ini:24]: Really confused about this update... Should still be 3.8? Is this a bug?

Comment (int3) [reply mypy-nofollow.ini:24]: uhm, did you see my explanatory comment above? :)

**Comment by Skylion007** [comment]
Apologies, the app didn't update. I still don't like the upping the mypy version because it could allow users to accidentally use APIs that aren't present in the minimum supported version. It also could mask some syntax errors. I'll let someone else chime in though, @malfet any thoughts?

**Comment by Skylion007** [comment]
The Any typing hack for versions hack doesn't seem like a huge issue. We already see behavior like for that conditional imports?

Another alternative is to run mypy with multiple versions through lintrunner. We already kind of do this for mypy-no follow etc so we could run 2 Python versions in lintrunner for min+max versions with mypy.

**Review by ezyang — APPROVED** [review]
I'm OK with this. We aren't really using the typechecking to catch people from using "too new" Python features, this has to be verified in CI anyway. If it simplifies the type annotations let's do it.

**Pull request closed without merging**